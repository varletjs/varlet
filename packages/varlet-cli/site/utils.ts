import { get } from 'lodash-es'
import { StyleProvider, type Theme, withSiteConfigNamespace } from '@varlet/cli/client'

export interface Menu {
  doc: string
  text: Record<string, string>
  type: MenuTypes
  useMobile?: boolean
}

export enum MenuTypes {
  TITLE = 1,
  COMPONENT = 2,
  DOCUMENTATION = 3,
}

export function removeEmpty(object: Record<string, string> = {}) {
  return Object.keys(object).reduce((record: Record<string, string>, key) => {
    const value = object[key]
    value && (record[key] = value)
    return record
  }, {})
}

export function isPhone() {
  return /Android|webOS|iPhone|iPod|BlackBerry|Pad/i.test(navigator.userAgent)
}

export function inIframe() {
  return window.self !== window.top
}

export function setTheme(config: Record<string, any>, name: Theme) {
  const styleVars = withSiteConfigNamespace(get(config, name, {}))

  StyleProvider(styleVars)
}

export function utoa(data: string): string {
  return btoa(unescape(encodeURIComponent(data)))
}
