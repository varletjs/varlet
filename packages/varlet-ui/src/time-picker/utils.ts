import { hours24, hoursAmpm } from './props'
import type { AmPm, AllowedTime } from './props'
import { toNumber } from '@varlet/shared'

type DisableProps = {
  time: number
  format: string
  ampm: AmPm | undefined
  hour: string
  minute: number
  max: string | undefined
  min: string | undefined
  disableHour: Array<string>
  allowedTime: AllowedTime | undefined
}

export const notConvert = (format: string, ampm: AmPm | undefined): boolean => format === '24hr' || ampm === 'am'

export const convertHour = (format: string, ampm: AmPm | undefined, hour: string) => {
  const index = hoursAmpm.findIndex((hourAmpm) => toNumber(hourAmpm) === toNumber(hour))
  const getHour = notConvert(format, ampm) ? hour : hours24[index]

  return {
    hourStr: getHour,
    hourNum: toNumber(getHour),
  }
}

export const getNumberTime = (time: string) => {
  const [hour, minute, second] = time.split(':')

  return {
    hour: toNumber(hour),
    minute: toNumber(minute),
    second: toNumber(second),
  }
}

export const getIsDisableMinute = (values: Omit<DisableProps, 'minute'>): boolean => {
  const { time, format, ampm, hour, max, min, disableHour } = values
  const { hourStr, hourNum } = convertHour(format, ampm, hour)
  let isBetweenMinMax = false
  let isAllow = false

  if (disableHour.includes(hourStr)) return true

  if (max && !min) {
    const { hour: maxHour, minute: maxMinute } = getNumberTime(max)
    isBetweenMinMax = maxHour === hourNum && time > maxMinute
  }

  if (!max && min) {
    const { hour: minHour, minute: minMinute } = getNumberTime(min)

    isBetweenMinMax = minHour === hourNum && time < minMinute
  }

  if (max && min) {
    const { hour: maxHour, minute: maxMinute } = getNumberTime(max)
    const { hour: minHour, minute: minMinute } = getNumberTime(min)

    isBetweenMinMax = (minHour === hourNum && time < minMinute) || (maxHour === hourNum && time > maxMinute)
  }

  if (values.allowedTime?.minutes) isAllow = values.allowedTime?.minutes(time)

  return isBetweenMinMax || isAllow
}

export const getIsDisableSecond = (values: DisableProps): boolean => {
  const { time, format, ampm, hour, minute, max, min, disableHour } = values
  const { hourStr, hourNum } = convertHour(format, ampm, hour)
  let isBetweenMinMax = false
  let isAllow = false

  if (disableHour.includes(hourStr)) return true

  if (max && !min) {
    const { hour: maxHour, minute: maxMinute, second: maxSecond } = getNumberTime(max)

    isBetweenMinMax = (maxHour === hourNum && maxMinute < minute) || (maxMinute === minute && time > maxSecond)
  }

  if (!max && min) {
    const { hour: minHour, minute: minMinute, second: minSecond } = getNumberTime(min)

    isBetweenMinMax = (minHour === hourNum && minMinute > minute) || (minMinute === minute && time > minSecond)
  }

  if (max && min) {
    const { hour: maxHour, minute: maxMinute, second: maxSecond } = getNumberTime(max)
    const { hour: minHour, minute: minMinute, second: minSecond } = getNumberTime(min)

    isBetweenMinMax =
      (maxHour === hourNum && maxMinute < minute) ||
      (minHour === hourNum && minMinute > minute) ||
      (maxHour === hourNum && maxMinute === minute && time > maxSecond) ||
      (minHour === hourNum && minMinute === minute && time < minSecond)
  }

  if (values.allowedTime?.seconds) isAllow = values.allowedTime?.seconds(time)

  return isBetweenMinMax || isAllow
}
