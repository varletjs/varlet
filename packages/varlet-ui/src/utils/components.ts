import {
  createApp,
  h,
  getCurrentInstance,
  inject,
  onUnmounted,
  computed,
  provide,
  reactive,
  isVNode,
  onMounted,
  onBeforeUnmount,
  nextTick,
  ref,
  onActivated,
  onDeactivated,
} from 'vue'
import type { Component, VNode, ComputedRef, ComponentInternalInstance, Ref, ComponentPublicInstance } from 'vue'
import { isArray, removeItem } from '@varlet/shared'

export interface MountInstance {
  instance: ComponentPublicInstance
  unmount: () => void
}

export interface ChildrenCounter {
  collect(instance: ComponentInternalInstance): void

  clear(instance: ComponentInternalInstance): void

  instances: ComponentInternalInstance[]
}

export interface BaseParentProvider<C> {
  collect(childProvider: C): void

  clear(childProvider: C): void
}

export function pickProps<T, U extends keyof T>(props: T, propsKey: U): T[U]
export function pickProps<T, U extends keyof T>(props: T, propsKey: U[]): Pick<T, U>
export function pickProps(props: any, propsKey: any): any {
  return Array.isArray(propsKey)
    ? propsKey.reduce((pickedProps: any, key) => {
        pickedProps[key] = props[key]
        return pickedProps
      }, {})
    : props[propsKey]
}

export function mount(component: Component): MountInstance {
  const app = createApp(component)
  const host = document.createElement('div')
  document.body.appendChild(host)

  return {
    instance: app.mount(host),
    unmount() {
      app.unmount()
      document.body.removeChild(host)
    },
  }
}

export function mountInstance(
  component: any,
  props: Record<string, any> = {},
  eventListener: Record<string, any> = {}
): {
  unmountInstance: () => void
} {
  const Host = {
    setup() {
      return () =>
        h(component, {
          ...props,
          ...eventListener,
        })
    },
  }

  const { unmount } = mount(Host)
  return { unmountInstance: unmount }
}

export function flatVNodes(subTree: any) {
  const vNodes: VNode[] = []

  const flat = (subTree: any) => {
    if (subTree?.component) {
      flat(subTree?.component.subTree)
      return
    }

    if (Array.isArray(subTree?.children)) {
      subTree.children.forEach((child: any) => {
        if (isVNode(child)) {
          vNodes.push(child)

          flat(child)
        }
      })
    }
  }

  flat(subTree)

  return vNodes
}

export function useAtChildrenCounter(key: symbol) {
  const instances: ComponentInternalInstance[] = reactive([])
  const parentInstance: ComponentInternalInstance = getCurrentInstance() as ComponentInternalInstance

  const sortInstances = () => {
    const vNodes: any[] = flatVNodes(parentInstance.subTree)

    instances.sort((a, b) => {
      return vNodes.indexOf(a.vnode) - vNodes.indexOf(b.vnode)
    })
  }

  const collect = (instance: ComponentInternalInstance) => {
    instances.push(instance)
    sortInstances()
  }

  const clear = (instance: ComponentInternalInstance) => {
    removeItem(instances, instance)
  }

  provide<ChildrenCounter>(key, {
    collect,
    clear,
    instances,
  })

  const length: ComputedRef<number> = computed(() => instances.length)

  return {
    length,
  }
}

export function useAtParentIndex(key: symbol) {
  if (!keyInProvides(key)) {
    return { index: null }
  }

  const childrenCounter: ChildrenCounter = inject<ChildrenCounter>(key) as ChildrenCounter

  const { collect, clear, instances } = childrenCounter

  const instance: ComponentInternalInstance = getCurrentInstance() as ComponentInternalInstance

  onMounted(() => {
    nextTick().then(() => collect(instance))
  })
  onUnmounted(() => {
    nextTick().then(() => clear(instance))
  })

  const index = computed(() => instances.indexOf(instance))

  return {
    index,
  }
}

export function useChildren<P, C>(key: symbol) {
  const childProviders: C[] = []

  const collect = (childProvider: C) => {
    childProviders.push(childProvider)
  }

  const clear = (childProvider: C) => {
    removeItem(childProviders, childProvider)
  }

  const bindChildren = (parentProvider: P) => {
    provide<P & BaseParentProvider<C>>(key, {
      collect,
      clear,
      ...parentProvider,
    })
  }

  return {
    childProviders,
    bindChildren,
  }
}

export function useParent<P, C>(key: symbol) {
  if (!keyInProvides(key)) {
    return {
      parentProvider: null,
      bindParent: null,
    }
  }

  const rawParentProvider = inject<P & BaseParentProvider<C>>(key) as P & BaseParentProvider<C>

  const { collect, clear, ...parentProvider } = rawParentProvider

  const bindParent = (childProvider: C) => {
    onMounted(() => collect(childProvider))
    onBeforeUnmount(() => clear(childProvider))
  }

  return {
    parentProvider,
    bindParent,
  }
}

export function keyInProvides(key: symbol) {
  const instance = getCurrentInstance() as any

  return key in instance.provides
}

export function useValidation() {
  const errorMessage: Ref<string> = ref('')

  const validate = async (rules: any, value: any, apis?: any): Promise<boolean> => {
    if (!isArray(rules) || !rules.length) {
      return true
    }

    const resArr = await Promise.all(rules.map((rule) => rule(value, apis)))

    return !resArr.some((res) => {
      if (res !== true) {
        errorMessage.value = String(res)
        return true
      }

      return false
    })
  }

  const resetValidation = () => {
    errorMessage.value = ''
  }

  const validateWithTrigger = async <T>(validateTrigger: T[], trigger: T, rules: any, value: any, apis?: any) => {
    if (validateTrigger.includes(trigger)) {
      ;(await validate(rules, value, apis)) && (errorMessage.value = '')
    }
  }

  return {
    errorMessage,
    validate,
    resetValidation,
    validateWithTrigger,
  }
}

export function useRouteListener(cb: () => void) {
  onMounted(() => {
    window.addEventListener('hashchange', cb)
    window.addEventListener('popstate', cb)
  })
  onUnmounted(() => {
    window.removeEventListener('hashchange', cb)
    window.removeEventListener('popstate', cb)
  })
}

export function useTeleport() {
  const disabled: Ref<boolean> = ref(false)

  onActivated(() => {
    disabled.value = false
  })

  onDeactivated(() => {
    disabled.value = true
  })

  return {
    disabled,
  }
}

export function exposeApis<T = Record<string, any>>(apis: T) {
  const instance = getCurrentInstance()
  if (instance) {
    Object.assign(instance.proxy!, apis)
  }
}

type ClassName = string | undefined | null
type Classes = (ClassName | [any, ClassName, ClassName?])[]

export function createNamespace(name: string) {
  const namespace = `var-${name}`

  const createBEM = (suffix?: string): string => {
    if (!suffix) return namespace

    return suffix.startsWith('--') ? `${namespace}${suffix}` : `${namespace}__${suffix}`
  }

  const classes = (...classes: Classes): any[] => {
    return classes.map((className) => {
      if (isArray(className)) {
        const [condition, truthy, falsy = null] = className
        return condition ? truthy : falsy
      }

      return className
    })
  }

  return {
    n: createBEM,
    classes,
  }
}

export function call<F extends (...arg: any) => any, P extends Parameters<F>>(
  fn?: F | null,
  ...arg: P
): ReturnType<F> | undefined {
  if (fn) return fn(...arg)
}
