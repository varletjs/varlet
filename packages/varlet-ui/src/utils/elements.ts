import { isNumber, isObject, isString, kebabCase, toNumber } from '@varlet/shared'
import { getGlobalThis } from './shared'
import type { StyleVars } from '../style-provider'

export function getLeft(element: HTMLElement): number {
  const { left } = element.getBoundingClientRect()

  return left + (document.body.scrollLeft || document.documentElement.scrollLeft)
}

export function getTop(element: HTMLElement): number {
  const { top } = element.getBoundingClientRect()

  return top + (document.body.scrollTop || document.documentElement.scrollTop)
}

export function getScrollTop(element: Element | Window): number {
  const top = 'scrollTop' in element ? element.scrollTop : element.pageYOffset

  // iOS scroll bounce cause minus scrollTop
  return Math.max(top, 0)
}

export function getScrollLeft(element: Element | Window): number {
  const left = 'scrollLeft' in element ? element.scrollLeft : element.pageXOffset

  return Math.max(left, 0)
}

export async function inViewport(element: HTMLElement): Promise<boolean> {
  await doubleRaf()
  const { top, bottom, left, right } = element.getBoundingClientRect()
  const { innerWidth, innerHeight } = window

  const xInViewport = left <= innerWidth && right >= 0
  const yInViewport = top <= innerHeight && bottom >= 0

  return xInViewport && yInViewport
}

export function getTranslate(el: HTMLElement) {
  const { transform } = window.getComputedStyle(el)
  return +transform.slice(transform.lastIndexOf(',') + 2, transform.length - 1)
}

export function getParentScroller(el: HTMLElement): HTMLElement | Window {
  let element = el

  while (element) {
    if (!element.parentNode) {
      break
    }

    element = element.parentNode as HTMLElement

    if (element === document.body || element === document.documentElement) {
      break
    }

    const scrollRE = /(scroll|auto)/
    const { overflowY, overflow } = window.getComputedStyle(element)

    if (scrollRE.test(overflowY) || scrollRE.test(overflow)) {
      return element
    }
  }

  return window
}

export function getAllParentScroller(el: HTMLElement): Array<HTMLElement | Window> {
  const allParentScroller: Array<HTMLElement | Window> = []
  let element: HTMLElement | Window = el

  while (element !== window) {
    element = getParentScroller(element as HTMLElement)
    allParentScroller.push(element)
  }

  return allParentScroller
}

export function getTarget(target: string | HTMLElement, componentName: string) {
  if (isString(target)) {
    const el = document.querySelector(target)

    if (!el) {
      throw Error(`[Varlet] ${componentName}: target element cannot found`)
    }

    return el as HTMLElement
  }

  if (isObject(target)) return target

  throw Error(`[Varlet] ${componentName}: type of prop "target" should be a selector or an element object`)
}

// example 1rem
export const isRem = (value: unknown): value is string => isString(value) && value.endsWith('rem')

// e.g. 1 || 1px
export const isPx = (value: unknown): value is string | number =>
  (isString(value) && value.endsWith('px')) || isNumber(value)

// e.g. 1%
export const isPercent = (value: unknown): value is string => isString(value) && value.endsWith('%')

// e.g. 1vw
export const isVw = (value: unknown): value is string => isString(value) && value.endsWith('vw')

// e.g. 1vh
export const isVh = (value: unknown): value is string => isString(value) && value.endsWith('vh')

// e.g. calc(1px + 1px)
export const isCalc = (value: unknown): value is string => isString(value) && value.startsWith('calc(')

// e.g. var(--color-primary)
export const isVar = (value: unknown): value is string => isString(value) && value.startsWith('var(')

// e.g. return 1
export const toPxNum = (value: unknown): number => {
  if (isNumber(value)) {
    return value
  }

  if (isPx(value)) {
    return +(value as string).replace('px', '')
  }

  if (isVw(value)) {
    return (+(value as string).replace('vw', '') * window.innerWidth) / 100
  }

  if (isVh(value)) {
    return (+(value as string).replace('vh', '') * window.innerHeight) / 100
  }

  if (isRem(value)) {
    const num = +(value as string).replace('rem', '')
    const rootFontSize = window.getComputedStyle(document.documentElement).fontSize

    return num * parseFloat(rootFontSize)
  }

  if (isString(value)) {
    return toNumber(value)
  }

  // % and other
  return 0
}

// e.g. return 1px 1% 1vw 1vh 1rem null var(--color-primary) calc(1px + 1px)
export const toSizeUnit = (value: unknown) => {
  if (value == null) {
    return undefined
  }

  if (isPercent(value) || isVw(value) || isVh(value) || isRem(value) || isCalc(value) || isVar(value)) {
    return value
  }

  return `${toPxNum(value)}px`
}

export const multiplySizeUnit = (value: unknown, quantity = 1) => {
  if (value == null) {
    return undefined
  }

  const legalSize = toSizeUnit(value) as string

  const unit = legalSize.match(/(vh|%|rem|px|vw)$/)![0]

  return `${parseFloat(legalSize) * quantity}${unit}`
}

export function requestAnimationFrame(fn: FrameRequestCallback): number {
  const globalThis = getGlobalThis()

  return globalThis.requestAnimationFrame ? globalThis.requestAnimationFrame(fn) : globalThis.setTimeout(fn, 16)
}

export function cancelAnimationFrame(handle: number): void {
  const globalThis = getGlobalThis()

  globalThis.cancelAnimationFrame ? globalThis.cancelAnimationFrame(handle) : globalThis.clearTimeout(handle)
}

export function nextTickFrame(fn: FrameRequestCallback) {
  requestAnimationFrame(() => {
    requestAnimationFrame(fn)
  })
}

export function doubleRaf() {
  return new Promise((resolve) => {
    requestAnimationFrame(() => {
      requestAnimationFrame(resolve)
    })
  })
}

export function raf() {
  return new Promise((resolve) => {
    requestAnimationFrame(resolve)
  })
}

interface ScrollToOptions {
  top?: number
  left?: number
  duration?: number
  animation: (progress: number) => number
}

export function scrollTo(
  element: HTMLElement | Window,
  { top = 0, left = 0, duration = 300, animation }: ScrollToOptions
): Promise<void> {
  const startTime = Date.now()

  const scrollTop = getScrollTop(element)
  const scrollLeft = getScrollLeft(element)

  return new Promise((resolve) => {
    const frame = () => {
      const progress = (Date.now() - startTime) / duration

      if (progress < 1) {
        const nextTop = scrollTop + (top - scrollTop) * animation(progress)
        const nextLeft = scrollLeft + (left - scrollLeft) * animation(progress)

        element.scrollTo(nextLeft, nextTop)
        requestAnimationFrame(frame)
      } else {
        element.scrollTo(left, top)
        resolve()
      }
    }

    requestAnimationFrame(frame)
  })
}

export function formatStyleVars(styleVars: StyleVars | null) {
  return Object.entries(styleVars ?? {}).reduce((styles, [key, value]) => {
    const cssVar = key.startsWith('--') ? key : `--${kebabCase(key)}`
    styles[cssVar] = value

    return styles
  }, {} as StyleVars)
}

export function supportTouch() {
  const inBrowser = typeof window !== 'undefined'
  return inBrowser && 'ontouchstart' in window
}

export function padStartFlex(style: string | undefined) {
  return style === 'start' || style === 'end' ? `flex-${style}` : style
}
