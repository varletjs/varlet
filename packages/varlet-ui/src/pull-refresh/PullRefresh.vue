<template>
  <div ref="freshNode" :class="n()" @touchstart="touchStart" @touchend="touchEnd" @touchcancel="touchEnd">
    <div
      ref="controlNode"
      :class="classes(n('control'), n('$-elevation--2'), [isSuccess, n('control-success')])"
      :style="controlStyle"
    >
      <var-icon
        :name="iconName"
        :transition="ICON_TRANSITION"
        :class="classes(n('icon'), [refreshStatus === 'loading' && iconHasChanged, n('animation')])"
        var-pull-refresh-cover
      />
    </div>
    <slot />
  </div>
</template>

<script lang="ts">
import VarIcon from '../icon'
import { defineComponent, ref, computed, watch, nextTick, type Ref } from 'vue'
import { getParentScroller, getScrollTop, getTarget } from '../utils/elements'
import { props, type RefreshStatus } from './props'
import { isString, toNumber } from '@varlet/shared'
import { call, createNamespace } from '../utils/components'
import { useEventListener, useMounted } from '@varlet/use'

const { n, classes } = createNamespace('pull-refresh')

const ICON_TRANSITION = 150

export default defineComponent({
  name: 'VarPullRefresh',
  components: {
    VarIcon,
  },
  props,
  setup(props) {
    const controlPosition: Ref<number> = ref(0)
    const freshNode: Ref<HTMLElement | null> = ref(null)
    const controlNode: Ref<HTMLElement | null> = ref(null)
    const startPosition: Ref<number> = ref(0)
    const distance: Ref<number | string> = ref('-125%')
    const iconName: Ref<string> = ref('arrow-down')
    const refreshStatus: Ref<RefreshStatus> = ref('default')
    const isEnd: Ref<boolean> = ref(false)

    let scroller: HTMLElement | Window
    let eventTargetScroller: HTMLElement | Window | null
    let changing: Promise<void>
    let startY = 0
    let deltaY = 0

    // https://github.com/varletjs/varlet/issues/509
    const iconHasChanged: Ref<boolean> = ref(true)

    const isTouchable = computed(
      () => refreshStatus.value !== 'loading' && refreshStatus.value !== 'success' && !props.disabled
    )

    const controlStyle = computed(() => ({
      transform: `translate3d(0px, ${
        isString(distance.value) ? distance.value : `${distance.value}px`
      }, 0px) translate(-50%, 0)`,
      transition: isEnd.value ? `transform ${props.animationDuration}ms` : undefined,
      background: isSuccess.value ? props.successBgColor : props.bgColor,
      color: isSuccess.value ? props.successColor : props.color,
    }))

    const maxDistance = computed(() => Math.abs(2 * controlPosition.value))

    const isSuccess = computed(() => refreshStatus.value === 'success')

    const changeIcon = (): Promise<void> => {
      return new Promise((resolve) => {
        window.setTimeout(() => {
          iconHasChanged.value = true
          resolve()
        }, ICON_TRANSITION)
      })
    }

    const lockEvent = (action: 'add' | 'remove') => {
      const el = 'classList' in scroller ? scroller : document.body

      el.classList[action](`${n()}--lock`)
    }

    const touchStart = (event: TouchEvent) => {
      if (controlPosition.value === 0) {
        const { width } = (controlNode.value as HTMLElement).getBoundingClientRect()
        controlPosition.value = -(width + width * 0.25)
      }

      startY = event.touches[0].clientY
      deltaY = 0
      eventTargetScroller = getParentScroller(event.target as HTMLElement)
    }

    const touchMove = (event: TouchEvent) => {
      if (!isTouchable.value) {
        return
      }

      if (eventTargetScroller !== scroller && getScrollTop(eventTargetScroller) > 0) {
        return
      }

      const scrollTop = getScrollTop(scroller)
      if (scrollTop > 0) {
        return
      }

      const isReachTop = scrollTop === 0
      const touch = event.touches[0]
      deltaY = touch.clientY - startY

      if (isReachTop && deltaY >= 0) {
        event.preventDefault()
      }

      if (refreshStatus.value !== 'pulling') {
        refreshStatus.value = 'pulling'
        startPosition.value = event.touches[0].clientY
      }

      if (isReachTop && distance.value > controlPosition.value) {
        lockEvent('add')
      }

      const moveDistance = (event.touches[0].clientY - startPosition.value) / 2 + controlPosition.value

      distance.value = moveDistance >= maxDistance.value ? maxDistance.value : moveDistance

      if (distance.value >= maxDistance.value * 0.2) {
        iconHasChanged.value = false
        iconName.value = 'refresh'
        changing = changeIcon()
      } else {
        iconName.value = 'arrow-down'
      }
    }

    const touchEnd = async () => {
      if (!isTouchable.value) return

      isEnd.value = true

      if (distance.value >= maxDistance.value * 0.2) {
        await changing
        refreshStatus.value = 'loading'
        distance.value = maxDistance.value * 0.3

        call(props['onUpdate:modelValue'], true)
        nextTick(() => {
          call(props.onRefresh)
        })
        lockEvent('remove')
      } else {
        refreshStatus.value = 'loosing'
        iconName.value = 'arrow-down'
        distance.value = controlPosition.value

        setTimeout(() => {
          isEnd.value = false
          lockEvent('remove')
        }, toNumber(props.animationDuration))
      }

      eventTargetScroller = null
    }

    const setScroller = () => {
      scroller = props.target ? getTarget(props.target, 'PullRefresh') : getParentScroller(freshNode.value!)
    }

    const reset = () => {
      setTimeout(() => {
        refreshStatus.value = 'default'
        iconName.value = 'arrow-down'
        isEnd.value = false
      }, toNumber(props.animationDuration))
    }

    watch(
      () => props.modelValue,
      (newValue) => {
        if (newValue === false) {
          isEnd.value = true
          refreshStatus.value = 'success'
          iconName.value = 'checkbox-marked-circle'
          setTimeout(() => {
            distance.value = controlPosition.value
            reset()
          }, toNumber(props.successDuration))
        }
      }
    )

    useMounted(setScroller)

    useEventListener(freshNode, 'touchmove', touchMove)

    return {
      n,
      classes,
      iconHasChanged,
      ICON_TRANSITION,
      refreshStatus,
      freshNode,
      controlNode,
      touchStart,
      touchMove,
      touchEnd,
      iconName,
      controlStyle,
      isSuccess,
    }
  },
})
</script>

<style lang="less">
@import '../styles/common';
@import '../styles/elevation';
@import '../icon/icon';
@import './pullRefresh';
</style>
