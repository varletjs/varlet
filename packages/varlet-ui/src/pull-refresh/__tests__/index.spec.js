import example from '../example'
import PullRefresh from '..'
import VarPullRefresh from '../PullRefresh'
import { mount } from '@vue/test-utils'
import { createApp } from 'vue'
import { delay, trigger } from '../../utils/jest'

test('test pullRefresh plugin', () => {
  const app = createApp({}).use(PullRefresh)
  expect(app.component(PullRefresh.name)).toBeTruthy()
})

test('test pullRefresh example', () => {
  const wrapper = mount(example, { attachTo: document.body })

  expect(wrapper.html()).toMatchSnapshot()
})

test('test the different states during the drop-down process', async () => {
  const template = `
  <var-pull-refresh
    bg-color="blue"
    color="red"
    success-color="purple"
    success-bg-color="#000"
  >
    <div style="height: 200px; width: 100%"></div>
  </var-pull-refresh>
  `

  const wrapper = mount({
    components: {
      [VarPullRefresh.name]: VarPullRefresh,
    },
    template,
  })

  const el = wrapper.find('.var-pull-refresh')

  await trigger(el, 'touchstart', 0, 0)
  await trigger(el, 'touchmove', 0, 50)
  await trigger(el, 'touchend', 0, 50)
  await delay(2000)

  expect(wrapper.html()).toMatchSnapshot()

  await trigger(el, 'touchmove', 0, 200)
  await delay(0)

  expect(wrapper.html()).toMatchSnapshot()

  await trigger(el, 'touchend', 0, 150)
  await delay(0)

  expect(wrapper.html()).toMatchSnapshot()
})

test('test disabled prop', async () => {
  const template = `
  <var-pull-refresh disabled >
    <div style="height: 200px; width: 100%"></div>
  </var-pull-refresh>
  `

  const wrapper = mount({
    components: {
      [VarPullRefresh.name]: VarPullRefresh,
    },
    template,
  })

  const el = wrapper.find('.var-pull-refresh')

  await trigger(el, 'touchstart', 0, 0)
  await trigger(el, 'touchmove', 0, 200)
  await delay(0)

  expect(wrapper.html()).toMatchSnapshot()
})

test('test pull-refresh event', async () => {
  const refresh = jest.fn()

  const template = `
  <var-pull-refresh v-model="isRefresh" @refresh="refresh" >
    <div style="height: 200px; width: 100%"></div>
  </var-pull-refresh>
  `

  const wrapper = mount({
    components: {
      [VarPullRefresh.name]: VarPullRefresh,
    },
    data() {
      return {
        isRefresh: false,
      }
    },
    methods: {
      refresh,
    },
    template,
  })

  const el = wrapper.find('.var-pull-refresh')

  await trigger(el, 'touchstart', 0, 0)
  await trigger(el, 'touchmove', 0, 200)
  await trigger(el, 'touchend', 0, 150)

  await delay(200)

  expect(wrapper.vm.isRefresh).toBe(true)
  expect(refresh).toHaveBeenCalledTimes(1)
  wrapper.vm.isRefresh = false

  await delay(3000)

  expect(wrapper.html()).toMatchSnapshot()
})
