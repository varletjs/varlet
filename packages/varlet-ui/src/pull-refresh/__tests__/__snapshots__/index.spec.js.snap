// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test disabled prop 1`] = `
"<div class=\\"var-pull-refresh\\">
  <div class=\\"var-pull-refresh__control var-elevation--2\\" style=\\"transform: translate3d(0px, -50px, 0px) translate(-50%, 0);\\"><i class=\\"var-icon var-icon--set var-icon-arrow-down var-pull-refresh__icon\\" style=\\"transition: transform 150ms;\\" var-pull-refresh-cover=\\"\\"></i></div>
  <div style=\\"height: 200px; width: 100%;\\"></div>
</div>"
`;

exports[`test pull-refresh event 1`] = `
"<div class=\\"var-pull-refresh\\">
  <div class=\\"var-pull-refresh__control var-elevation--2\\" style=\\"transform: translate3d(0px, -50px, 0px) translate(-50%, 0);\\"><i class=\\"var-icon var-icon--set var-icon-arrow-down var-pull-refresh__icon\\" style=\\"transition: transform 150ms;\\" var-pull-refresh-cover=\\"\\"></i></div>
  <div style=\\"height: 200px; width: 100%;\\"></div>
</div>"
`;

exports[`test pullRefresh example 1`] = `
"<div class=\\"var-pull-refresh\\">
  <div class=\\"var-pull-refresh__control var-elevation--2\\" style=\\"transform: translate3d(0px, -50px, 0px) translate(-50%, 0);\\"><i class=\\"var-icon var-icon--set var-icon-arrow-down var-pull-refresh__icon\\" style=\\"transition: transform 150ms;\\" var-pull-refresh-cover=\\"\\"></i></div>
  <div class=\\"var-cell var-cell--border\\">
    <!--v-if-->
    <div class=\\"var-cell__content\\">
      <div class=\\"var-cell__title\\">List Item 1</div>
      <!--v-if-->
    </div>
    <!--v-if-->
  </div>
  <div class=\\"var-cell var-cell--border\\">
    <!--v-if-->
    <div class=\\"var-cell__content\\">
      <div class=\\"var-cell__title\\">List Item 2</div>
      <!--v-if-->
    </div>
    <!--v-if-->
  </div>
  <div class=\\"var-cell var-cell--border\\">
    <!--v-if-->
    <div class=\\"var-cell__content\\">
      <div class=\\"var-cell__title\\">List Item 3</div>
      <!--v-if-->
    </div>
    <!--v-if-->
  </div>
  <div class=\\"var-cell var-cell--border\\">
    <!--v-if-->
    <div class=\\"var-cell__content\\">
      <div class=\\"var-cell__title\\">List Item 4</div>
      <!--v-if-->
    </div>
    <!--v-if-->
  </div>
  <div class=\\"var-cell var-cell--border\\">
    <!--v-if-->
    <div class=\\"var-cell__content\\">
      <div class=\\"var-cell__title\\">List Item 5</div>
      <!--v-if-->
    </div>
    <!--v-if-->
  </div>
  <div class=\\"var-cell var-cell--border\\">
    <!--v-if-->
    <div class=\\"var-cell__content\\">
      <div class=\\"var-cell__title\\">List Item 6</div>
      <!--v-if-->
    </div>
    <!--v-if-->
  </div>
  <div class=\\"var-cell var-cell--border\\">
    <!--v-if-->
    <div class=\\"var-cell__content\\">
      <div class=\\"var-cell__title\\">List Item 7</div>
      <!--v-if-->
    </div>
    <!--v-if-->
  </div>
  <div class=\\"var-cell var-cell--border\\">
    <!--v-if-->
    <div class=\\"var-cell__content\\">
      <div class=\\"var-cell__title\\">List Item 8</div>
      <!--v-if-->
    </div>
    <!--v-if-->
  </div>
  <div class=\\"var-cell var-cell--border\\">
    <!--v-if-->
    <div class=\\"var-cell__content\\">
      <div class=\\"var-cell__title\\">List Item 9</div>
      <!--v-if-->
    </div>
    <!--v-if-->
  </div>
  <div class=\\"var-cell var-cell--border\\">
    <!--v-if-->
    <div class=\\"var-cell__content\\">
      <div class=\\"var-cell__title\\">List Item 10</div>
      <!--v-if-->
    </div>
    <!--v-if-->
  </div>
</div>"
`;

exports[`test the different states during the drop-down process 1`] = `
"<div class=\\"var-pull-refresh\\">
  <div class=\\"var-pull-refresh__control var-elevation--2\\" style=\\"transform: translate3d(0px, -50px, 0px) translate(-50%, 0); color: purple; background: rgb(0, 0, 0);\\"><i class=\\"var-icon var-icon--set var-icon-arrow-down var-pull-refresh__icon\\" style=\\"transition: transform 150ms;\\" var-pull-refresh-cover=\\"\\"></i></div>
  <div style=\\"height: 200px; width: 100%;\\"></div>
</div>"
`;

exports[`test the different states during the drop-down process 2`] = `
"<div class=\\"var-pull-refresh\\">
  <div class=\\"var-pull-refresh__control var-elevation--2\\" style=\\"transform: translate3d(0px, 50px, 0px) translate(-50%, 0); color: purple; background: rgb(0, 0, 0);\\"><i class=\\"var-icon var-icon--set var-icon-arrow-down var-icon--shrinking var-pull-refresh__icon\\" style=\\"transition: transform 150ms;\\" var-pull-refresh-cover=\\"\\"></i></div>
  <div style=\\"height: 200px; width: 100%;\\"></div>
</div>"
`;

exports[`test the different states during the drop-down process 3`] = `
"<div class=\\"var-pull-refresh\\">
  <div class=\\"var-pull-refresh__control var-elevation--2\\" style=\\"transform: translate3d(0px, 50px, 0px) translate(-50%, 0); color: purple; transition: transform 300ms; background: rgb(0, 0, 0);\\"><i class=\\"var-icon var-icon--set var-icon-arrow-down var-icon--shrinking var-pull-refresh__icon\\" style=\\"transition: transform 150ms;\\" var-pull-refresh-cover=\\"\\"></i></div>
  <div style=\\"height: 200px; width: 100%;\\"></div>
</div>"
`;
